zerver/views/auth.py

```diff
@@ -589,7 +589,10 @@ def dev_direct_login(request: HttpRequest, **kwargs: Any) -> HttpResponse:
     if user_profile is None:
         return HttpResponseRedirect(reverse('dev_not_supported'))
     do_login(request, user_profile)
-    return HttpResponseRedirect(user_profile.realm.uri)
+
+    next = request.GET.get('next', '')
+
+    return HttpResponseRedirect(user_profile.realm.uri + next)
```

Comment:
This should use `urlparse.urljoin`, I think.  

Also, you need a security check to make sure that `next` is a safe URL (also, do you need to URL-decode it or something if there's interesting characters in it?).  I think you can adapt your solution from this:

https://www.reddit.com/r/django/comments/32s9ag/how_do_i_set_the_next_context_variable_in_django/cqe6kcl/

(certainly the security check there is right), except that we need to specify the specific subdomain, so a relative URL doesn't make sense for us.

Reply:
Probably makes sense to extract a function for doing the `next` handling logic on login, and then we can test that with a bunch of interesting cases (e.g. leading `/` on URL, no leading `/` on URL, etc.)

Reply:
I think we don't need to url decode because that's handled by django.
```
if isinstance(query_string, bytes):
            # query_string normally contains URL-encoded data, a subset of ASCII.
            try:
                query_string = query_string.decode(encoding)
            except UnicodeDecodeError:
                # ... but some user agents are misbehaving :-(
                query_string = query_string.decode('iso-8859-1')
        for key, value in limited_parse_qsl(query_string, **parse_qsl_kwargs):
```
Above snippet is from django Request class 
https://docs.djangoproject.com/en/2.0/ref/request-response/#django.http.QueryDict.__init__
https://docs.djangoproject.com/en/2.0/_modules/django/http/request/#QueryDict.__init__

Reply:
Addressed the url safe check and then redirect issue :)